src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -8380,6 +8380,48 @@ internal static string HandleDoubleAndSingleQuote(ref string wordToComplete)
             return quote;
         }
 
+        /// <summary>
+        /// Enumerates quoted and unquoted completion text.
+        /// This makes it easier to handle different variations of completions with consideration of quotes.
+        /// </summary>
+        /// <param name="wordToComplete">The word to complete.</param>
+        /// <param name="possibleCompletionValues">The possible completion values to iterate.</param>
+        /// <param name="toolTipMapping">The optional tool tip mapping delegate.</param>
+        /// <param name="resultType">The optional completion result type. Default is Text.</param>
+        /// <returns></returns>
+        internal static IEnumerable<CompletionResult> EnumerateQuotedAndUnquotedCompletionText(
+            string wordToComplete,
+            IEnumerable<string> possibleCompletionValues,
+            Func<string, string> toolTipMapping = null,
+            CompletionResultType resultType = CompletionResultType.Text)
+        {
+            string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
+            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
+
+            foreach (string value in possibleCompletionValues)
+            {
+                if (pattern.IsMatch(value))
+                {
+                    string completionText = quote == string.Empty
```

Comment:
Since this is a general purpose function you should probably add some logic to escape special characters in the value.

Reply:
Hmm can probably do `Regex.Escape(value)` to make sure that the strings inputted in are all escaped?

Reply:
@MartinGC64 Good catch. If there is no existing utility method which does this already I'll add that logic in.

Reply:
Wonder if i can just use `Regex.Escape` here. Given we are just removing special characters for wildcard matching. Although I don't see much usage in existing code so wondering if I am missing something here?

Reply:
I also think it would be worth parametrizing it with `bool escape` so it can be turned off incase for some reason we need to keep preserved string without escaping.

Reply:
If the special characters are single and double quotas what are examples we could use for tests? 
I suggest to postpone this to keep the PR small.

Reply:
It makes sense to delay this. I only find this useful if the input strings are not guaranteed to have special chars but currently none of the completers changed here have strings with special chars so perhaps we revisit when we refactor a completer that has special chars? It's probably also easier to review that in a separate PR.

Reply:
I don't see any reason to include an option to disable the escaping. The only reason you'd want that is if you plan on doing additional processing and then escape them properly, but the method is generating complete completion results that cannot be modified anyway so that's hardly a concern.
As for how to implement it, you can take a look at the various suggestions in this thread: https://github.com/PowerShell/PowerShell/issues/24493

Reply:
@MartinGC94 Can you have a look at a297ab849d35f86952aa3c2592d488cb09843d6b. I have included code to do escaping on special chars, let me know if I'm missing anything.

Reply:
There's a lot more problematic chars to handle, see this switch I use for escaping paths: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L5003 some of them are unique to the path handling but many of them are not.  
To keep things simple I'd just use `IndexOfAny` to check for problematic characters: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L5226 and if any are found, force single quotes and escape with: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/lang/codegen.cs#L21

Reply:
@MartinGC94 Thanks. Let me know if a24641669c4580d2f6ae8b6778b0ee7a4dc06ad6 looks right. I will add some tests later if I get some time.